app/src/org/commcare/android/nfc/NfcWriteActivity.java

```diff
@@ -0,0 +1,103 @@
+ * consumed by this activity. CommCare will then decide what type of NdefRecord to write to the tag
+ * based upon the user-provided 'type' argument:
+ *
+ *    - If type is 'text', CommCare will write a record of type NdefRecord.RTD_TEXT
+ *    - If the type is anything else, CommCare will assume the user is attempting to write a
+ *    custom/external record type (which will be qualified by the domain argument)
+ *
+ * If an error occurs during the write action, an appropriate error toast will be shown.
+ *
+ * @author Aliza Stone
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+public class NfcWriteActivity extends NfcActivity {
+
+    public static final String NFC_PAYLOAD_TO_WRITE = "payload";
+
+    private String payloadToWrite;
+    private String typeForPayload;
+
+    @Override
+    protected void initFields() {
+        super.initFields();
+        this.payloadToWrite = getIntent().getStringExtra(NFC_PAYLOAD_TO_WRITE);
+        this.typeForPayload = getIntent().getStringExtra(NFC_PAYLOAD_SINGLE_TYPE_ARG);
+    }
+
+    @Override
+    protected boolean requiredFieldsMissing() {
+        if (this.payloadToWrite == null || this.payloadToWrite.equals("")) {
+            finishWithErrorToast("nfc.write.no.payload");
+            return true;
+        } else if (this.typeForPayload == null || this.typeForPayload.equals("")) {
+            finishWithErrorToast("nfc.write.no.type");
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    protected void dispatchActionOnTag(Tag tag) {
+        writeMessageToNfcTag(tag);
+    }
+
+    private void writeMessageToNfcTag(Tag tag) {
+        Ndef ndefObject = Ndef.get(tag);
+        try {
+            NdefRecord record = NdefRecordUtil.createNdefRecord(this.typeForPayload,
+                    this.domainForType, this.payloadToWrite);
+            NdefMessage msg = new NdefMessage(new NdefRecord[]{record});
+            ndefObject.connect();
+            ndefObject.writeNdefMessage(msg);
+            ndefObject.close();
+            finishWithToast("nfc.write.success", true);
+        } catch (IOException e) {
+            finishWithErrorToast("nfc.write.io.error");
+        } catch (FormatException e) {
+            finishWithErrorToast("nfc.write.msg.malformed");
+        } catch (IllegalArgumentException e) {
+            finishWithErrorToast("nfc.write.type.not.supported");
+        }
+    }
+
+    @Override
+    protected void setResultExtrasBundle(Intent i, boolean success) {
+        Bundle responses = new Bundle();
+        responses.putString("nfc_write_result", success ? "success" : "failure");
+        i.putExtra(IntentCallout.INTENT_RESULT_EXTRAS_BUNDLE, responses);
+    }
+
+    @Override
+    protected void setResultValue(Intent i, boolean success) {
+        if (success) {
+            i.putExtra(IntentCallout.INTENT_RESULT_VALUE, "Wrote value: " + payloadToWrite);
```

Comment:
Why are you appending `Wrote value: ` here? That seems dangerous/wrong (IE if I were parsing this intent response I would not expect that header)

Reply:
Yea you're right, will remove.